gift-certificates-app/dao/src/main/java/com/epam/esm/dao/impl/CertificateDAOImpl.java

```diff
@@ -1,94 +1,116 @@
 package com.epam.esm.dao.impl;
 
 import com.epam.esm.bean.Certificate;
-import com.epam.esm.bean.OrderDTO;
-import com.epam.esm.bean.SearchDTO;
+import com.epam.esm.bean.Search;
+import com.epam.esm.bean.Sort;
+import com.epam.esm.bean.Tag;
 import com.epam.esm.dao.CertificateDAO;
-import com.epam.esm.dao.mapper.CertificateMapper;
-import com.epam.esm.dao.util.CertificateGetSQLRequest;
-import com.epam.esm.dao.util.CertificateUpdateParameters;
-import com.epam.esm.dao.util.CertificateUpdateSQLRequest;
+import com.epam.esm.dao.TagDAO;
+import com.epam.esm.exception.ResourceAlreadyExistsException;
+import com.epam.esm.exception.ResourceHasLinksException;
 import com.epam.esm.exception.ResourceNotFoundException;
+import org.springframework.beans.BeanUtils;
+import org.springframework.beans.BeanWrapper;
+import org.springframework.beans.BeanWrapperImpl;
 import org.springframework.beans.factory.annotation.Autowired;
-import org.springframework.jdbc.core.JdbcTemplate;
-import org.springframework.jdbc.support.GeneratedKeyHolder;
-import org.springframework.jdbc.support.KeyHolder;
 import org.springframework.stereotype.Repository;
+import org.springframework.transaction.annotation.Transactional;
+import org.springframework.util.CollectionUtils;
 
-import java.sql.PreparedStatement;
-import java.sql.Timestamp;
-import java.time.Instant;
-import java.time.temporal.ChronoUnit;
+import javax.persistence.EntityManager;
+import javax.persistence.PersistenceContext;
+import javax.persistence.PersistenceException;
+import java.util.HashSet;
 import java.util.List;
+import java.util.Optional;
+import java.util.Set;
 
```

Comment:
documentation comments are needed

Reply:
Public APIs should be documented.  I documented Service. Maybe this is enough?